model_zoo/pytorch/nanogpt/train.py

```diff
@@ -27,65 +27,50 @@
 from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 from torch.distributed.fsdp.wrap import size_based_auto_wrap_policy
 from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.utils.data import DataLoader, Dataset
+
+os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
 
 local_rank = None
 master_process = False
 
 
-def get_data_loader(data_dir):
-    train_data = np.memmap(
-        os.path.join(data_dir, "train.bin"), dtype=np.uint16, mode="r"
-    )
-    val_data = np.memmap(
-        os.path.join(data_dir, "val.bin"), dtype=np.uint16, mode="r"
-    )
-    # Attempt to derive vocab_size from the dataset
-    meta_path = os.path.join(data_dir, "meta.pkl")
-    meta_vocab_size = None
-    if os.path.exists(meta_path):
-        with open(meta_path, "rb") as f:
-            meta = pickle.load(f)
-    meta_vocab_size = meta["vocab_size"]
-    print(f"found vocab_size = {meta_vocab_size} (inside {meta_path})")
-    return train_data, val_data, meta_vocab_size
+class GPTDataset(Dataset):
+    def __init__(self, data_path, block_size=128):
+        self.data = np.memmap(data_path, dtype=np.uint16, mode="r")
+        self.block_size = block_size
 
+    def __len__(self):
+        return len(self.data) - self.block_size
 
-def get_batch(
-    split,
-    train_data,
-    val_data,
-    device_type="cpu",
-    device="cpu",
+    def __getitem__(self, idx):
+        x = torch.from_numpy(
+            self.data[idx : idx + self.block_size].astype(np.int64) 
+        )  # noqa E203 
+        y = torch.from_numpy(
+            self.data[idx + 1 : idx + 1 + self.block_size].astype(np.int64)   
+        )  # noqa E203
+        return x, y
+
+
+def get_data_loaders(
+    data_dir,
     batch_size=12,
     block_size=128,
+    device_type="cpu",
+    device="cpu",
+    use_fsdp="True",
 ):
-    data = train_data if split == "train" else val_data
-    ix = torch.randint(len(data) - block_size, (batch_size,))
-    x = torch.stack(
-        [
-            torch.from_numpy(
-                (data[i : i + block_size]).astype(np.int64)  # noqa: E203 E501
-            )
-            for i in ix
-        ]
-    )
-    y = torch.stack(
-        [
-            torch.from_numpy(
-                (data[i + 1 : i + 1 + block_size]).astype(np.int64)  # noqa
-            )
-            for i in ix
-        ]
+    train_dataset = GPTDataset(os.path.join(data_dir, "train.bin"), block_size)
+    val_dataset = GPTDataset(os.path.join(data_dir, "val.bin"), block_size)
+    with open(os.path.join(data_dir, "meta.pkl"), "rb") as f:
+        meta = pickle.load(f)
+    train_loader = DataLoader(
+        train_dataset, batch_size=batch_size, shuffle=True
     )
-    if device_type == "cuda":
-        # Pin arrays x,y, which allows us to move them
-        # to GPU asynchronously (non_blocking=True)
-        x, y = x.pin_memory().to(device, non_blocking=True), y.pin_memory().to(
```

Comment:
Can use `pin_memory` in the dataloader?

Reply:
Yes, it was my oversight, and I will add this implementation.